FirebaseDatabaseSwift/Sources/Codable/DataSnapshot+ReadDecodable.swift

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseDatabase
+
+extension Database.Decoder {
+  public static var defaultDecoder: () -> Database.Decoder = {
+    .init()
+  }
+}
+
+extension DataSnapshot {
+  /// Retrieves the value of a snapshot and converts it to an instance of
+  /// caller-specified type.
+  /// Throws `DecodingError.valueNotFound`
+  /// if the document does not exist and `T` is not an `Optional`.
+  ///
+  /// See `Database.Decoder` for more details about the decoding process.
+  ///
+  /// - Parameters
+  ///   - type: The type to convert the document fields to.
+  ///   - decoder: The decoder to use to convert the document. Defaults to use
+  ///              default decoder.
+  public func data<T: Decodable>(as type: T.Type,
+                                 decoder: Database.Decoder = Database.Decoder
+                                   .defaultDecoder()) throws -> T {
+    try decoder.decode(T.self, from: value ?? NSNull())
+  }
```

Comment:
Hi Ryan,
I'll naturally go with your recommendations, but I just have a small comment to your discussion:
There's no need for using `snapshot.exists`. The user is in full control about whether or not the expect existance/optionality.
For instance, if you have a snapshot of some `Entity` that may or may not exist, then you can write:
```
snapshot.data(as: Entity?.self)
```
While if you have some `Configuration` that must exist (ie. you consider it an error for it to be missing), then you can do:
```
snapshot.data(as: Configuration.self)
```
And the final example where you have snapshots from `childAdded`. Here you know for a fact that data can exist and can parse it without optionality.

So my main point is that I believe that the user has more power in the choice of adding optionality to the generic parameter themselves rather than have it 'forced'. And I don't see the need for calling `snapshot.exists` - you just imply optionality when decoding.

Reply:
Another thought I just had:
When adding the `Combine` API to get some `Publisher` from a `DatabaseReference`, I would think that the API was similar to `data(as:)`. Would you for that API also expect that '.decodedEntityPublisher(as: Configuration.self)' would give a publisher of optional `Configuration` rather than allowing the user to specify the optionality in the generic parameter here too?

Reply:
Is there any possibility that my comments here may be reconsidered? :-)
I implemented the other two suggested changes.

Reply:
Hi @ryanwilson ,
I haven't been able to create mock instances of `DataSnapshot` in Swift. The last time I checked I couldn't find any pointers in the codebase as to how I might do it.
Do you have any suggestions here?